### PR TITLE
[WIP] Initial commit of feature benchmarks 

### PR DESCRIPTION
**Pull request by akrzos**
Feature benchmarks will benchmark portions of the Miq code with live providers of different sizes.  The goal is to produce the latency required to complete specific tasks in a controlled and repeatable manner.

This first commit benchmarks:
* Discovery 
  * Providers
* Refresh
  * Provider
  * Targeted Host
  * Targeted VM 
* Refresh with Broker for Vmware providers
  * Provider
  * Targeted Host
  * Targeted VM
* Capacity and Utilization
  * Realtime Capture on VM
  * Realtime Capture on Host
* Events
  * Measures Processing time for PowerOn/PowerOff chained events
  * Measures Processing time for two "non-event_handler" events

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit b68852fb2b3abac01796234430a7331bdad3452b:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./cfme/tests/perf/feature/test_capacity_and_utilization.py:26:21: E127 continuation line over-indented for visual indent
```
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit a85cdd571c55d2baeac80c1b7c8ce953632e0431:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit a1a373f143fb61d5a3dea10f90f0e69cd738ba02:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 71e5702e7f4bf5e0645567a2e2990d5e5326d5d8:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

File: utils/perf_feature_stats.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        logger.info('Appending to: features-statistics.csv')
+        outputfile = csv_file_path.open('a', ensure=True)
+    else:
+        logger.info('Writing to: features-statistics.csv')
+        outputfile = csv_file_path.open('w', ensure=True)
+        outputfile.write('feature,test,provider,runs,minimum,average,median,maximum,stddev,90th'
+            ',99th\n')
+
+    outputfile.write('{},{},{},{},{},{},{},{},{},{},{}\n'.format(feature, test, provider,
+        len(timings), minimum, average, median, maximum, stddev, percentile90, percentile99))
```
Comment (mfalesni) [inline utils/perf_feature_stats.py]: Hmm, maybe https://docs.python.org/2/library/csv.html ?

**Comment by dajoRH** [comment]
Lint report for commit 4d7bff23855718c95228695846a61dcaddbfdd4d:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

File: utils/perf_feature_stats.py (hunk 1 above)
Comment (psav) [inline utils/perf_feature_stats.py]: Or even just 
```"{}\n".format(",".join([feature, test, provider, len(timings), minimum, average, median, maximum, stddev, percentile90, percentile99]))```

File: utils/perf_feature_stats.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    outputfile.close()
+
+    logger.info('Stats (min/avg/med/max/stddev/90/99) : ' + str(minimum).rjust(6) + '/' +
+        str(average).rjust(6) + '/' + str(median).rjust(6) + '/' + str(maximum).rjust(6) + '/' +
+        str(stddev).rjust(6) + '/' + str(percentile90).rjust(6) + '/' + str(percentile99).rjust(6))
```
Comment (psav) [inline utils/perf_feature_stats.py]: ```'/'.join([str(a).rjust(6) for a in [minimum, average..........]])```

File: utils/perf_feature_stats.py (hunk 1 above)
Comment (mfalesni) [inline utils/perf_feature_stats.py]: But that still leaves door open for escape-related things to sneak in. I know it is probably not happening here but we have a standard library here so why not use it :)

Comment (psav) [inline utils/perf_feature_stats.py]: Agreed :)

**Comment by dajoRH** [comment]
Lint report for commit 58e6975fd037ae2fe31d71d3aa542fa01fc5e35c:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 6432a5f1044cc1163177adafc92321ec87759eb0:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./cfme/tests/perf/feature/test_capacity_and_utilization.py:4:1: F401 'logger' imported but unused
```
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 4ee27208148dc1d29bd3d562b3bad8bcc94bb146:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 7fc15b27ab5f8ad3a400666e8dbf1c15d55afb69:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit fa059042eaa2792bf6dfc9f70b615e6dfe0bd63e:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 74e07a3f3d5d57729e6eda95319164bdf550feab:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 1c83c5534b07c0f9c4d31b07f52d99dd1db302e2:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 27b350cb124a566604b867d052bd5978dad8f126:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 2506df30dfc97b1a65673a5b8bd93dda61c8c3f5:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./cfme/tests/perf/feature/test_event.py:287:101: E501 line too long (191 > 100 characters)
./cfme/tests/perf/feature/test_event.py:289:101: E501 line too long (122 > 100 characters)
./cfme/tests/perf/feature/test_event.py:293:101: E501 line too long (126 > 100 characters)
./cfme/tests/perf/feature/test_event.py:296:32: W291 trailing whitespace
```
:trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 532f72d3bb74100ca3f11d5d686acec2f4d652c1:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit fa1dc553877853dd7ff2816b31d2852d336348a6:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 2c39bbaf2deac7326258bb44ef317a4f3dcd8d60:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 231e27d8cb506e25aa6a6affc15e2357117162e8:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./conftest.py:8:1: F401 'PIPE' imported but unused
./conftest.py:8:1: F401 'Popen' imported but unused
./conftest.py:11:1: F401 're' imported but unused
```
:trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 2bfa2171c3f33b7ee0a0ecc098835e03d2f6f538:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 79bd4544f471411e446ca46c97cd020e54225519:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 8de2b4a44ba7ba196e3226129103d9d3bc4ae80f:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 15fc5739ec9db7e3529029270e12c129ed80917f:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 2a3abf6851cd4f5b78bb74d24f17ec8658e1a3c1:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./artifactor/plugins/post_result.py:13:1: F401 'os' imported but unused
./artifactor/plugins/post_result.py:18:1: F401 'read_env' imported but unused
```
:trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 26ce4967e75a644f757d405ef2c151d928783df2:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit c915a9c67991048b9a3a6380961ffdb68ef818bd:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit ed7715bef5d05b029f9f072964594a60ee63dcaf:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit e900574779e2e5f7681f1e74d9202e4ad4ab844b:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit b164b1dc18a6f2e1d4a15e717067609dea02142d:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 6ca9ba8dbe6321e525ae5342532a7bb67525bc28:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./cfme/tests/perf/feature/test_refresh_broker.py:34:101: E501 line too long (116 > 100 characters)
```
:trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 8527f398ce41dc1f3f56ef816b8394f498138fef:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./cfme/tests/perf/feature/test_refresh_broker.py:66:17: E127 continuation line over-indented for visual indent
./cfme/tests/perf/feature/test_refresh_broker.py:94:17: E127 continuation line over-indented for visual indent
./cfme/tests/perf/feature/test_refresh_broker.py:134:5: F841 local variable 'reps' is assigned to but never used
./cfme/tests/perf/feature/test_refresh_broker.py:144:17: E127 continuation line over-indented for visual indent
./cfme/tests/perf/feature/test_refresh_broker.py:162:5: F841 local variable 'reps' is assigned to but never used
./cfme/tests/perf/feature/test_refresh_broker.py:171:17: E127 continuation line over-indented for visual indent
./cfme/tests/perf/feature/test_refresh.py:61:17: E127 continuation line over-indented for visual indent
./cfme/tests/perf/feature/test_refresh.py:87:17: E127 continuation line over-indented for visual indent
./cfme/tests/perf/feature/test_refresh.py:125:5: F841 local variable 'reps' is assigned to but never used
./cfme/tests/perf/feature/test_refresh.py:133:101: E501 line too long (104 > 100 characters)
./cfme/tests/perf/feature/test_refresh.py:151:5: F841 local variable 'reps' is assigned to but never used
./cfme/tests/perf/feature/test_refresh.py:158:101: E501 line too long (109 > 100 characters)
```
:trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit e858bae8452b39c33ad8850d9e5d4b425da763a0:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit f5b0b9529af08672e7b499bfa2037dc435e48996:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by akrzos** [comment]
Closing, old and outdated, newer branch with latest benchmarks in my repo.

**Pull request closed without merging**